core/src/main/scala/kafka/log/OffsetIndex.scala

```diff
@@ -218,24 +219,34 @@ object OffsetIndex extends Logging {
   */
 class LazyOffsetIndex(@volatile private var _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true) {
   @volatile private var offsetIndex: Option[OffsetIndex] = None
+  private val lock = new ReentrantLock()
 
   def file: File = {
-    if (offsetIndex.isDefined)
-      offsetIndex.get.file
-    else
-      _file
+    inLock(lock) {
+      if (offsetIndex.isDefined)
+        offsetIndex.get.file
+      else
+        _file
+    }
   }
 
   def file_=(f: File): Unit = {
-    if (offsetIndex.isDefined)
-      offsetIndex.get.file = f
-    else
-      _file = f
+    inLock(lock) {
+      if (offsetIndex.isDefined)
+        offsetIndex.get.file = f
+      else
+        _file = f
+    }
   }
 
   def get: OffsetIndex = {
-    if (offsetIndex.isEmpty)
-      offsetIndex = Some(new OffsetIndex(_file, baseOffset, maxIndexSize, writable))
+    if (offsetIndex.isEmpty) {
+      inLock(lock) {
+        if (offsetIndex.isEmpty) {
+          offsetIndex = Some(new OffsetIndex(_file, baseOffset, maxIndexSize, writable))
+        }
+      }
+    }
     offsetIndex.get
```

Comment:
The usual pattern for something like this is to read to a local var first. And avoid another volatile read for the common case.

Reply:
@ijuma good catch, fixed